cpp/src/arrow/python/arrow_to_pandas.cc

```diff
@@ -347,8 +346,17 @@ class PandasBlock {
     } else {
       block_dims[0] = num_rows_;
     }
-    RETURN_NOT_OK(PyArray_NewFromPool(ndim, block_dims, npy_type,
-                                      /*arrow_type=*/nullptr, options_.pool, &block_arr));
+    PyArray_Descr* descr = internal::GetSafeNumPyDtype(npy_type);
+    if (PyDataType_REFCHK(descr)) {
+      // ARROW-6876: if the array has refcounted items, let Numpy
+      // own the array memory so as to decref elements on array destruction
+      block_arr = PyArray_SimpleNewFromDescr(ndim, block_dims, descr);
```

Comment:
Another option would be to have a buffer that decrefs the pointers in its dtor, but that seems like a bit too much work here